src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,427 @@
+/-
+Copyright (c) 2021 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Eric Wieser
+-/
+
+import ring_theory.ideal.basic
+import ring_theory.ideal.operations
+import linear_algebra.finsupp
+import ring_theory.graded_algebra.basic
+
+/-!
+
+# Homogeneous ideal of a graded algebra
+
+This file defines homogeneous ideals of `graded_algebra A` where `A : ι → ideal R`and operations on
+them:
+* `mul`, `inf`, `Inf` of homogeneous ideals are homogeneous;
+* `⊤`, `⊥`, i.e. the trivial ring and `R` are homogeneous;
+* `radical` of a homogeneous ideal is homogeneous.
+-/
+
+open set_like direct_sum set
+open_locale big_operators pointwise direct_sum
+
+section homogeneous_def
+
+variables {ι R A : Type*} [comm_ring R] [comm_ring A] [algebra R A]
+variables (𝒜 : ι → submodule R A)
+variables [decidable_eq ι] [add_comm_monoid ι]  [graded_algebra 𝒜]
+variable (I : ideal A)
+
+/--An `I : ideal R` is homogeneous if for every `r ∈ I`, all homogeneous components
+  of `r` are in `I`.-/
+def ideal.is_homogeneous : Prop :=
+∀ (i : ι) ⦃r : A⦄, r ∈ I → (graded_algebra.decompose 𝒜 r i : A) ∈ I
+
+/--For any `comm_ring R`, we collect the homogeneous ideals of `R` into a type.-/
+abbreviation homogeneous_ideal : Type* := { I : ideal A // I.is_homogeneous 𝒜 }
+
+end homogeneous_def
+
+section homogeneous_core
+
+variables {ι R A : Type*} [comm_ring R] [comm_ring A] [algebra R A]
+variables (𝒜 : ι → submodule R A)
+variable (I : ideal A)
+
+/-- For any `I : ideal R`, not necessarily homogeneous, `I.homogeneous_core' 𝒜`
+is the largest homogeneous ideal of `R` contained in `I`. -/
+def ideal.homogeneous_core' : ideal A :=
+ideal.span (coe '' ((coe : subtype (is_homogeneous 𝒜) → A) ⁻¹' I))
+
+lemma ideal.homogeneous_core'_is_mono : monotone (ideal.homogeneous_core' 𝒜) :=
+λ I J I_le_J, ideal.span_mono $ set.image_subset _ $ λ x, @I_le_J _
+
+lemma ideal.homogeneous_core'_le_ideal : I.homogeneous_core' 𝒜 ≤ I :=
+ideal.span_le.2 $ image_preimage_subset _ _
+
+end homogeneous_core
+
+section is_homogeneous_ideal_defs
+
+variables {ι R A : Type*} [comm_ring R] [comm_ring A] [algebra R A]
+variables (𝒜 : ι → submodule R A)
+variables [decidable_eq ι] [add_comm_monoid ι]  [graded_algebra 𝒜]
+variable (I : ideal A)
+
+lemma ideal.is_homogeneous_iff_forall_subset :
+  I.is_homogeneous 𝒜 ↔ ∀ i, (I : set A) ⊆ graded_algebra.proj 𝒜 i ⁻¹' I :=
+iff.rfl
+
+lemma ideal.is_homogeneous_iff_subset_Inter :
+  I.is_homogeneous 𝒜 ↔ (I : set A) ⊆ ⋂ i, graded_algebra.proj 𝒜 i ⁻¹' ↑I :=
+subset_Inter_iff.symm
+
+lemma ideal.is_homogeneous.exists_iff_eq_span :
+  (∃ (S : set (homogeneous_submonoid 𝒜)), I = ideal.span (coe '' S)) ↔
+    I = I.homogeneous_core' 𝒜 :=
+(set.image_preimage.compose (submodule.gi _ _).gc).exists_eq_l _
```

Comment:
I think you should get rid of this lemma and inline it below - the name is misleading because the statement doesn't once mention `ideal.is_homogeneous`. See below...